src/plugins/expressions/common/expression_types/filter.ts

```diff
@@ -34,9 +35,35 @@ export interface Filter {
   query?: string | null;
 }
 
+const migrator = {
+  isUnsupported: (p: any) => !p.version || p.version < 2,
+  isLegacy: (p: any) => p.version && p.version < 3,
+  migrate: (legacyFilter: any) => ({ ...legacyFilter, version: 3 }),
+};
+
+function parseAndValidate(s: string): Filter {
+  try {
+    const parsed = JSON.parse(s);
+    if (migrator.isUnsupported(parsed)) {
+      throw new Error('LEGACY');
+    }
+    return parsed;
+  } catch (e) {
+    if (e.message === 'LEGACY') {
+      throw new Error('Provided filter string is malformed or no longer supported');
+    }
+    throw new Error('Unable to parse filter string.');
+  }
+}
+
 export const filter = (): ExpressionType<typeof name, Filter> => ({
   name,
   from: {
+    string: s => {
+      const { isLegacy, migrate } = migrator;
+      const parsed = parseAndValidate(s);
+      return isLegacy(parsed) ? migrate(parsed) : parsed;
+    },
```

Comment:
The whole point here is that the "owner" of filters can set up the casting/migrations without the author of the `combineFilters` function needing to know anything: they just receive the correct, up-to-date filter shape as an object inside their function body as long as they specify their arg as a `filter` type.

Reply:
Also worth pointing out that @streamich has [floated the idea](https://github.com/elastic/kibana/issues/55167#issuecomment-576942164) of making type converters pluggable by third parties in some future version of the expressions service:

```js
registerTypeConverter('string', 'filter', (string) => { /*...*/ })
```

Reply:
I think this could be built on top of the general migration system, but my point was, who is writing `migrate: (legacyFilter: any) => ({ ...legacyFilter, version: 3 }),` function and what is the implementation?

I think it would work if it used the generic persistableStateFactory we talked about.  Code could look something like

```
 from: {
    string: s => {
    try {
      const { type, ...state } = JSON.parse(s);
      const persistableStateFactory = persistableStateFactories.get(type);

      if (!persistableStateFactory) {
        throw new Error('NOT SUPPORTED');
      }

      if (persistableStateFactory.isDeprecated()) {
        console.warn('LEGACY - upgrade soon to avoid breakage!');
      }
      
      let migratedState = state;
      while (persistableStateFactory.isDeprecated()) {
        const { type: newType, state: migratedState } = persistableStateFactory.migrate(state);
        persistableStateFactory = persistableStateFactories.get(newType);
      }
       return migratedState;
    } catch (e) {
      throw new Error('Unable to parse filter string.');
    }
  }  
}
```

I was thinking `persistableStateFactory.create(state)` would return an instance of whatever the factory was for, but since canvas doesn't deal with instances, in this case you'd probably just return the migrated state.